File: Algoritimos/Busca/BinarySearch/BinarySearch.java
package Busca.BinarySearch;
import java.util.Scanner;

public class BinarySearch {

    public static void Search(int arr[], int x, int esq, int dir) {
        int meio = (esq + dir) / 2;
        if (esq > dir) {
            System.out.println("Número não encontrado.");
            return;
        }
        if (arr[meio] == x) {
            System.out.println("O numero " + x + " esta no indice " + meio);
            return;
        }
        if (x > arr[meio]) {
            esq = meio + 1;
            Search(arr, x, esq, dir);
        } else if (x < arr[meio]) {
            dir = meio - 1;
            Search(arr, x, esq, dir);
        }

    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int arr[] = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
        int dir = arr.length - 1;
        int esq = 0;

        int x = sc.nextInt();

        Search(arr, x, esq, dir);

        sc.close();
    }
}

// Dividir e conquistar
// Condição de parada é esq < dir
// esq = meio + 1 || dir = meio - 1 (para chamada recursiva)

File: Algoritimos/Busca/BinarySearch/BinarySearch.c
#include <stdio.h>

int buscaBinaria(int arr[], int esq, int dir, int x){
    if(esq>dir){
        return -1;
    }
    int meio = esq + (dir-esq)/2;
    if (arr[meio] == x){
        return meio;
    }
    if (x > arr[meio]){
        return buscaBinaria(arr, meio+1, dir, x);
    }
    return buscaBinaria(arr, esq, meio-1, x);
}

int main(){
    int a[] = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
    int len = sizeof(a)/sizeof(a[0]);

    int x;
    scanf ("%d", &x);

    int busca = buscaBinaria(a, 0, len-1, x);

    if (busca==-1){
        printf("%d nao foi encontrado", x);
    }
    else{
        printf ("%d foi encontrado em %d", x, busca);
    }
    return 0;
}
